specs/ServerCertificate.md

```diff
@@ -0,0 +1,264 @@
+Server Certificate API
+===
+# Background
+
+The WebView2 team has been asked for an API to intercept when WebView2 cannot verify server's digital certificate while loading a web page.
+This API provides an option to trust the server's TLS certificate at the application level and render the page without prompting the user about the TLS error or can cancel the request.
+
+# Description
+
+We propose adding `ReceivingServerCertificateError` API that allows you to verify TLS certificate,  and either continue the request
+to load the resource or cancel the request.
+
+When the event is raised, WebView2 will pass a `CoreWebView2ReceivingServerCertificateErrorEventArgs` , which lets you view the TLS certificate request Uri, error information, inspect the certificate metadata and several options for responding to the TLS request.
+
+* You can perform your own verification of the certificate and allow the navigation to proceed if you trust it.
+* You can choose to cancel the request.
+* You can choose to display the default TLS interstitial page to let user respond to the request.
+
+We also propose adding a `ClearServerCertificateErrorOverrideCache` API that clears certificate decisions that were added in response to proceeding with TLS certificate errors.
+
+# Examples
+
+## Win32 C++
+``` cpp
+// This example bypasses the default TLS interstitial page using the ReceivingServerCertificateError
+// event handler and continues with the navigation to a server with a TLS certificate that is signed by
+// an authority that WebView2 doesn't trust. Otherwise, cancel the request.
+void SettingsComponent::EnableCustomServerCertificateError()
+{
+  if (m_webview)
+  {
+    if (m_ServerCertificateErrorToken.value == 0)
+    {
+         CHECK_FAILURE(m_webview->add_ReceivingServerCertificateError(
+             Callback<ICoreWebView2ReceivingServerCertificateErrorEventHandler>(
+             [this](
+               ICoreWebView2* sender,
+               ICoreWebView2ReceivingServerCertificateErrorEventArgs* args) {
+               COREWEBVIEW2_WEB_ERROR_STATUS errorStatus;
+               CHECK_FAILURE(args->get_ErrorStatus(&errorStatus));
+
+               wil::com_ptr<ICoreWebView2Certificate> certificate = nullptr;
+               CHECK_FAILURE(args->get_ServerCertificate(&certificate));
+
+               // Continues with the navigation to a server with a TLS certificate if
+               // the certificate is invalid and signed by badssl authority.
+               if (errorStatus == COREWEBVIEW2_WEB_ERROR_STATUS_CERTIFICATE_IS_INVALID &&
+                        (certificateAuthority.get() == L"*.badssl.com" ||
```

Comment:
A cert that isn't from a trusted root could put whatever it wants in the certificateAuthority field right? Is this how we expect our customers to use this, or is there a more secure way of doing this check? (Or if the secure way of doing this check is too big or complicated to include in this sample we can also change this to be a method call like ValidateCertificate and instead of showing the definition of ValidateCertificate add a big comment explaining what ValidateCertificate would be doing.)

Reply:
Added a function def `ValidateCertificate` with comment, a secure way of trusting the cert at application level.